fix(connections): add Prometheus labels for folders and devices (fixes #9519)

mwmiller opened the pull request:
### Purpose

This is intended to address concerns raised in #9519.  The intent is to add an info gauge (as suggested by @calmh) to ease human cross-referencing of stats as stored by IDs and meaningful strings. 

### Testing

I have run the automated tests.  I did not see where these tests exercise the metrics reporting in particular.
As such, I have not added any new tests.  An interested party could use their favorite metrics explorer to find the new gauges and correlate them with other labeled metrics.

### Documentation

This change will be visible to users in their metrics databases, but not within any interface of syncthing to my knowledge. I am happy to document the new metric gauges if there is a reasonable place to do so.

### Additional Info

This is my first attempt to contribute to `syncthing`.  Process errors or omissions are somewhat expected.  I am happy to iterate on this work and make such adjustments as help the project.

mwmiller committed:
Create an info gauge for folder_info

This includes the information most likely to make it possible for a
human to correlate the ids to something meaningful.

There were a couple of approaches possible, but adjusting
`metrics.go` seems like the least invasive and likely to cause
difficulty.  A function on `folder` would require expanding the
surface area of `prometheus` which feel worse than expanding
syncthing's own `config` type.

mwmiller committed:
Create a device_info gauge

This very much mirrors the approach in doing the folder_info gauge

rasa commented on lib/connections/metrics.go:
```diff
@@ -18,8 +20,30 @@ var metricDeviceActiveConnections = promauto.NewGaugeVec(prometheus.GaugeOpts{
+
+func registerDeviceInfoGauge(dc config.DeviceConfiguration) {
+	// Create a dynamic "info" gauge to help users
+	// map IDs to humane strings.
```
```suggestion
	// map IDs to human-readable strings.
```

rasa commented on lib/model/metrics.go:
```diff
@@ -99,3 +103,25 @@ func registerFolderMetrics(folderID string) {
+
+func registerInfoGauge(fc config.FolderConfiguration) {
+	// Create a dynamic "info" gauge to help users
+	// map IDs to humane strings.
```
```suggestion
	// map IDs to human-readable strings.
```

mwmiller committed:
Improve language slightly

- `human-readable` from `humane`
- `registerFolderInfoGauge` from `registerInfoGauge`

mwmiller commented:
Thanks, @rasa for the review.  I have incorporated your suggested changes into the latest rebased push.  I wasn't really sure how to express the sense properly. I also updated the folder function name for symmetry and searchability.

Also, thank you to @tomasz1986 for fixing the PR title to match conventions.  I saw that a fixing commit should have such in the short message, but my fix spanned 2 commits, so I floundered.

calmh commented on lib/connections/metrics.go:
```diff
@@ -18,8 +20,30 @@ var metricDeviceActiveConnections = promauto.NewGaugeVec(prometheus.GaugeOpts{
+	info_gauge := prometheus.NewGaugeFunc(
+		prometheus.GaugeOpts{
+			Namespace: "syncthing",
+			Name:      "device_info_" + did,
```
Generally one wouldn't do this as separate metrics, but one metric with multiple label values, e.g.

```
syncthing_device_info{id="aaa", name="bbb"} = 1
syncthing_device_info{id="ccc", name="ddd"} = 1
```

etc, otherwise it will be very difficult to use in joins I think. Probably the easiest way to do this would be to have a custom collector that has the config in entirety and generates the metrics when requested to

mwmiller replied:
Ok, cool, I will look into a custom collector.

Is it expected that most syncthing users will be using Grafana for observability monitoring?  As I progress, I would like to base my ergonomic/usability testing on whatever is most likely to benefit actual users.

calmh replied:
I'm not sure what you're asking really; I expect a very tiny minority of Syncthing users to use any form of observability. Of those that do, I think Grafana on either Prometheus or VictoriaMetrics etc is the absolutely most likely combination, yes.

mwmiller replied:
You've given me enough on which to work.  I want to recreate a "common" setup and confirm that my final code "works" ergonomically to expose the information intended.

Sorry for the delay in replying.  We had a death in the family which has kept me busy throughout this week.  I expect to be able to work on a custom collector early next week.
